### DISCUSSION TITLE
Adding custom sources as separate repos?

### DISCUSSION THREAD

**Discussion by ElPiloto** [nvim-neo-tree/neo-tree.nvim#403] [Internal Development]
One awkward thing I've bumped up against when trying to add my own source is having to modify `defaults.lua` and `setup/init.lua` (where [sources are defined](https://github.com/nvim-neo-tree/neo-tree.nvim/blob/9138180835c5ff61cbae6a3a28c025913621d138/lua/neo-tree/setup/init.lua#L14)).  If it weren't for needing to make changes to those files, I could actually develop my custom source in a new repository and just put my source files (init.lua, components.lua, commands.lua) in `my_repo_root/lua/neo-tree/sources/my_custom_source/` and the imports for `neo-tree.sources.$SOURCE` would work fine.  I'm not sure if this is an intended path for development (to be able to add sources as standalone repos), but it might be nice to have.  If desirable, adding something like telescope's `register_extension` to modify the sources table in `setup/init.lua` would probably do it.  Or `setup/init.lua` could take in sources via user config in the setup call.   What do you think?

**Comment by cseickel** [comment]
I think it's a good idea and one I have considered before. The only thing stopping me was:

1. Up until now no one has made a custom source.
2. When extensions are seperate from the main repo, you are going to have more of a problem with breaking changes and abandoned extensions.

All in all, it's probably best to decouple sources from the main repo like you suggested. I would say the thing to do is to have the extension register itself, it would really be a simple change to make.

**Reply by ElPiloto** [reply]
That would be great! Do you have the bandwidth to do that anytime soon? I managed to make a decent version of my custom source (for my job's version of workspaces), but haven't distributed it yet in hopes that we could have external sources in neo-tree.

**Reply by cseickel** [reply]
Yes. I want to finish up a couple of other small things I am working on first. If I had to guess, I'd say I'll get to this in a week or two.

**Reply by ElPiloto** [reply]
Great, I'll keep an eye out for that.  I really appreciate it.  Thank you.

**Comment by cseickel** [comment]
OK, I think I have a functional setup for this now, it's still sitting in a PR: https://github.com/nvim-neo-tree/neo-tree.nvim/pull/406

You can pull that branch and give it a try. Basically, you just need to add a `sources` list to your neo-tree config that includes the name of your module. Everything else is unchanged. You still put your config in the neo-tree setup call.

I extracted the `example` source out to a separate repo as an example: https://github.com/nvim-neo-tree/example-source

Let me know what you think.

**Reply by ElPiloto** [reply]
Somehow didn't get a notification for this. Will try it out ASAP

**Reply by ElPiloto** [reply]
This is looking great, thanks for the quick turnaround!

Here's a very simple, (incomplete) custom source I added for exposing ezbookmarks.nvim as a custom source: https://github.com/ElPiloto/neo-tree-ezbookmarks.nvim .

I might actually have to switch to another bookmarks backend as there seems to be one critical feature missing from that plugin, but I'm linking it here anyways just in case it's helpful in testing out this feature.

**Comment by prncss-xyz** [comment]
Hi, just created a source for neo-tree [here](https://github.com/prncss-xyz/neo-tree-zk.nvim). (An integration for zk.) I've been using it for a few days and so far I'm quite happy with it. (Congrats for the extensible design.) My only (small) issue is that I have not found a way to set a default config (would make user experience slightly simpler). Also where do we maintain a list of user sources? ([nvim-cmp](https://github.com/hrsh7th/nvim-cmp/wiki/List-of-sources) uses a wiki and an hashtag) If there are things you think I should do differently don't hesitate to tell. best regards!

**Reply by cseickel** [reply]
That's awesome, I'm glad to hear that it was easy to create!

Regarding default configs, I just merged a change to the `main` branch that would allow you to provide defaults by exporting 'default_config` as a key in your main module.

**Reply by cseickel** [reply]
As far as a list of sources, how about this? https://github.com/nvim-neo-tree/neo-tree.nvim/wiki/External-Sources  

I can link to that wiki page in the README.

For a tag (or actually github Topic), how about `neo-tree-source`?

**Reply by prncss-xyz** [reply]
Great, I've added the default_config; will wait that it's merged into branch 2.0 before updating the README.

As for the topic and wiki page that's all good to me.

**Comment by mrbjarksen** [comment]
I was just about to submit a pull request for a diagnostics source I've been tinkering with the last few weeks when I noticed this discussion. I reworked it a bit and placed it [here](https://github.com/mrbjarksen/neo-tree-diagnostics.nvim). It seems to work great. I did make a few minor changes to the internals, however, and might submit some pull requests.

One thing I think would be worth adding is a way to hook into the neo-tree highlight module, for consistency in highlight group definitions. Currently, I'm awkwardly falling back to already defined highlight groups, which makes it cumbersome to customize component highlights. I reckon it doesn't have to be more complicated than exposing `create_highlight_group`.

Well done, by the way, on designing this plugin! It's been a true delight to explore and work with.

**Reply by cseickel** [reply]
> I was just about to submit a pull request for a diagnostics source I've been tinkering with the last few weeks when I noticed this discussion. I reworked it a bit and placed it [here](https://github.com/mrbjarksen/neo-tree-diagnostics.nvim).

Cool! It just so happens I have been thinking about doing this but I decided it would be quicker to just pretty up my quickfix window with custom formatters for now. I'm glad someone else made it, I'm looking forward to trying it out.


> Well done, by the way, on designing this plugin! It's been a true delight to explore and work with.

Thanks!


> One thing I think would be worth adding is a way to hook into the neo-tree highlight module, for consistency in highlight group definitions. Currently, I'm awkwardly falling back to already defined highlight groups, which makes it cumbersome to customize component highlights. I reckon it doesn't have to be more complicated than exposing `create_highlight_group`.

I don't see why not. Feel free to submit a PR with this or other changes that are helpful in creating external sources.
